tap_github/client.py

```diff
@@ -0,0 +1,271 @@
+        "message": "User doesn't have permission to access the resource."
+    },
+    404: {
+        "raise_exception": NotFoundException,
+        "message": "The resource you have specified cannot be found. Alternatively the access_token is not valid for the resource"
+    },
+    409: {
+        "raise_exception": ConflictError,
+        "message": "The request could not be completed due to a conflict with the current state of the server."
+    },
+    422: {
+        "raise_exception": UnprocessableError,
+        "message": "The request was not able to process right now."
+    },
+    500: {
+        "raise_exception": InternalServerError,
+        "message": "An error has occurred at Github's end."
+    }
+}
+
+def raise_for_error(resp, source):
+    """
+    Retrieve the error code and the error message from the response and return custom exceptions accordingly.
+    """
+    error_code = resp.status_code
+    try:
+        response_json = resp.json()
+    except JSONDecodeError:
+        response_json = {}
+
+    if error_code == 404:
+        details = ERROR_CODE_EXCEPTION_MAPPING.get(error_code).get("message")
+        if source == "teams":
+            details += ' or it is a personal account repository'
+        message = "HTTP-error-code: 404, Error: {}. Please refer \'{}\' for more details.".format(details, response_json.get("documentation_url"))
+        LOGGER.warning(message)
+        # Don't raise a NotFoundException
+        return None
+
+    message = "HTTP-error-code: {}, Error: {}".format(
+        error_code, ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get("message", "Unknown Error") if response_json == {} else response_json)
+
+    exc = ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get("raise_exception", GithubException)
+    raise exc(message) from None
+
+def calculate_seconds(epoch):
+    """
+    Calculate the seconds to sleep before making a new request.
+    """
+    current = time.time()
+    return int(round((epoch - current), 0))
+
+def rate_throttling(response):
+    """
+    For rate limit errors, get the remaining time before retrying and calculate the time to sleep before making a new request.
+    """
+    if int(response.headers['X-RateLimit-Remaining']) == 0:
+        seconds_to_sleep = calculate_seconds(int(response.headers['X-RateLimit-Reset']))
+
+        if seconds_to_sleep > MAX_SLEEP_SECONDS:
+            message = "API rate limit exceeded, please try after {} seconds.".format(seconds_to_sleep)
+            raise RateLimitExceeded(message) from None
+
+        LOGGER.info("API rate limit exceeded. Tap will retry the data collection after %s seconds.", seconds_to_sleep)
+        time.sleep(seconds_to_sleep)
+
+class GithubClient:
+    """
+    The client class used for making REST calls to the Github API.
+    """
+    def __init__(self, config):
+        self.config = config
+        self.session = requests.Session()
+        self.verify_access_for_repo()
```

Comment:
@prijendev I would not recommend keeping a verifying check in the __init__() function because the repo access will be validated every time the sync would happen.
Previously, this check was happening only during discovery mode, which is expected. Reference - https://github.com/singer-io/tap-github/blob/4f7ba58e501bd5026979194a0c25f31d32d3ffe0/tap_github/__init__.py#L400

Reply:
I guess we should not remove access checking from here. As we got one card where if someone directly run the sync mode at that time we should pass the catalog and siply run the sync without selecting any stream. 
@prijendev If you are removing this line then you need to handle that scenario as well.

Reply:
@karanpanchal-crest Access chech before object creation is better as if we don't have access then no need to sync further.

Reply:
What is the final take on this comment @prijendev ?

Reply:
Tap will check credentials during discover mode and print a warning message only if the repo is not found. In the sync mode, we will give a warning message containing all invalid or not accessible repo names.